src/storage/data_table.cpp

```diff
@@ -174,60 +120,28 @@ bool DataTable::Update(const common::ManagedPointer<transaction::TransactionCont
-  // Assume block is full.
-  if (block_index == insertion_head_.load()) {
-    // If the header block is full, move the header to point to the next block.
-    insertion_head_++;
-  }
-
-  // If there are no more free blocks, create a new empty block and point the insertion_head to it.
-  if (insertion_head_.load() == blocks_.size()) {
-    RawBlock *new_block = NewBlock();
-    // Take the latch and insert the block. The insertion head will already have the right index.
-    common::SpinLatch::ScopedSpinLatch guard_block(&blocks_latch_);
-    blocks_.push_back(new_block);
-  }
-}
-
 TupleSlot DataTable::Insert(const common::ManagedPointer<transaction::TransactionContext> txn,
                             const ProjectedRow &redo) {
   TERRIER_ASSERT(redo.NumColumns() == accessor_.GetBlockLayout().NumColumns() - NUM_RESERVED_COLUMNS,
                  "The input buffer never changes the version pointer column, so it should have  exactly 1 fewer "
                  "attribute than the DataTable's layout.");
 
-  // Insertion header points to the first block that has free tuple slots
-  // Once a txn arrives, it will start from the insertion header to find the first
-  // idle (no other txn is trying to get tuple slots in that block) and non-full block.
-  // If no such block is found, the txn will create a new block.
-  // Before the txn writes to the block, it will set block status to busy.
-  // The first bit of block insert_head_ is used to indicate if the block is busy
-  // If the first bit is 1, it indicates one txn is writing to the block.
-
   TupleSlot result;
-  auto block_index = insertion_head_.load();
-  RawBlock *block;
-
+  uint64_t current_insert_idx = insert_index_.load();  // the index into which we will try to insert the tuple
+  RawBlock *block;                                     // the block into which the insert will occur
   while (true) {
     // No free block left
-    if (block_index == blocks_.size()) {
-      RawBlock *new_block = NewBlock();
-      TERRIER_ASSERT(accessor_.SetBlockBusyStatus(new_block), "Status of new block should not be busy");
-      // No need to flip the busy status bit
-      accessor_.Allocate(new_block, &result);
-      // take latch
-      common::SpinLatch::ScopedSpinLatch guard(&blocks_latch_);
-      // insert block
-      blocks_.push_back(new_block);
-      block = new_block;
-      break;
-    }
-
-    {
-      common::SpinLatch::ScopedSpinLatch guard(&blocks_latch_);
-      block = blocks_[block_index];
+    uint64_t size = blocks_size_;
+    if (current_insert_idx >= size) {
+      block = NewBlock();
+      common::SharedLatch::ScopedExclusiveLatch latch(&blocks_latch_);
+      blocks_.push_back(block);
+      current_insert_idx = blocks_.size() - 1;
+      blocks_size_ = blocks_.size();
+    } else {
+      common::SharedLatch::ScopedSharedLatch latch(&blocks_latch_);
+      block = const_cast<RawBlock *>(blocks_[current_insert_idx]);
```

Comment:
Why did you add `const_cast`?  As far as I can tell, it is unnecessary because neither the vector nor the lvalue are const qualified.

Reply:
done